web/lua/opmserver.lua

```diff
@@ -2049,6 +2081,351 @@ function _M.do_docs_page()
 end
 
 
+function _M.do_login_page()
+    local github_login_url = github_oauth.get_login_url()
+
+    view.show("login", {
+        github_login_url = github_login_url,
+    })
+end
+
+
+local function db_insert_user_session(user)
+    local u = user
+    local q = quote_sql_str
+    local sql
+
+    local ip = ngx_var.http_x_forwarded_for or ngx_var.remote_addr
+    local ua = ngx_var.http_user_agent
+    local token = gen_random_string()
+    sql = "insert into users_sessions (user_id, token, ip, user_agent) "
+          .. "values ("
+          .. q(u.id) .. ", "
+          .. q(token) .. ", "
+          .. q(ip) .. ", "
+          .. q(ua) .. ") returning id"
+
+    local res = query_db(sql)
+    local session_id = res[1].id
+
+    if not session_id then
+        return nil, "session_id not found"
+    end
+
+    return token
+end
+
+
+local function set_cookie(opmid, remember)
+    local ck = cookie_tab
+    ck.key = "OPMID"
+    ck.value = opmid
+    ck.max_age = nil
+    ck.httponly = true
+
+    if remember then
+        ck.max_age = LOGIN_COOKIE_MAX_TTL
+    end
+
+    local cookie = resty_cookie:new()
+    cookie:set(ck)
+end
+
+
+local function get_opmid()
+    local cookie = ngx_var.http_cookie
+    if not cookie then
+        return nil
+    end
+
+    local sid_name = 'OPMID'
+    local ssid, err = re_match(cookie, sid_name .. "=([\\da-z]{32})", "jo")
+    if err then
+        log.error("failed to find session id from cookie: ", err)
+        return nil
+    end
+
+    if not ssid then
+        return nil
+    end
+
+    return ssid[1]
+end
+
+
+local function get_current_user()
+    local ctx = ngx.ctx
+    if ctx.curr_user then
+        return ctx.curr_user
+    end
+
+    local token = get_opmid()
+    if not token then
+        return nil, "OPMID missing", true
+    end
+
+    local sql = "select * from users_sessions where token = "
+                .. quote_sql_str(token) .. " limit 1"
+    local res, err = query_db(sql)
+
+    if not res or #res == 0 then
+        return nil, "session not found"
```

Comment:
Better print the `err` (if it exists) to the error log for troubleshooting purposes.